source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,278 @@
+    }
+  }
+};
+
+void RedisCluster::startPreInit() {
+  for (const DnsDiscoveryResolveTargetPtr& target : dns_discovery_resolve_targets_) {
+    target->startResolve();
+  }
+}
+
+void RedisCluster::updateAllHosts(const Upstream::HostVector& hosts_added,
+                                  const Upstream::HostVector& hosts_removed,
+                                  uint32_t current_priority) {
+  Upstream::PriorityStateManager priority_state_manager(*this, local_info_, nullptr);
+
+  priority_state_manager.initializePriorityFor(redis_discovery_session_->locality_lb_endpoint_);
+  for (const Upstream::HostSharedPtr& host : redis_discovery_session_->hosts_) {
+    if (redis_discovery_session_->locality_lb_endpoint_.priority() == current_priority) {
+      priority_state_manager.registerHostForPriority(
+          host, redis_discovery_session_->locality_lb_endpoint_);
+    }
+  }
+
+  priority_state_manager.updateClusterPrioritySet(
+      current_priority, std::move(priority_state_manager.priorityState()[current_priority].first),
+      hosts_added, hosts_removed, absl::nullopt);
+}
+
+// DnsDiscoveryResolveTarget
+RedisCluster::DnsDiscoveryResolveTarget::DnsDiscoveryResolveTarget(
+    RedisCluster& parent, const std::string& dns_address, const uint32_t port,
+    const envoy::api::v2::endpoint::LocalityLbEndpoints& locality_lb_endpoint,
+    const envoy::api::v2::endpoint::LbEndpoint& lb_endpoint)
+    : parent_(parent), dns_address_(dns_address), port_(port),
+      locality_lb_endpoint_(locality_lb_endpoint), lb_endpoint_(lb_endpoint) {}
+
+RedisCluster::DnsDiscoveryResolveTarget::~DnsDiscoveryResolveTarget() {
+  if (active_query_) {
+    active_query_->cancel();
+  }
+}
+
+void RedisCluster::DnsDiscoveryResolveTarget::startResolve() {
+  ENVOY_LOG(trace, "starting async DNS resolution for {}", dns_address_);
+
+  active_query_ = parent_.dns_resolver_->resolve(
+      dns_address_, parent_.dns_lookup_family_,
+      [this](const std::list<Network::Address::InstanceConstSharedPtr>&& address_list) -> void {
+        active_query_ = nullptr;
+        ENVOY_LOG(trace, "async DNS resolution complete for {}", dns_address_);
+        parent_.redis_discovery_session_->registerDiscoveryAddress(address_list, port_);
+        parent_.redis_discovery_session_->startResolve();
+      });
+}
+
+// RedisCluster
+RedisCluster::RedisDiscoverySession::RedisDiscoverySession(
+    Envoy::Extensions::Clusters::Redis::RedisCluster& parent,
+    NetworkFilters::Common::Redis::Client::ClientFactory& client_factory)
+    : parent_(parent), dispatcher_(parent.dispatcher_),
+      resolve_timer_(parent.dispatcher_.createTimer([this]() -> void { startResolve(); })),
+      client_factory_(client_factory), buffer_timeout_(0) {}
+
+namespace {
+Network::Address::InstanceConstSharedPtr
+ProcessCluster(const NetworkFilters::Common::Redis::RespValue& value) {
+  ASSERT(value.type() == NetworkFilters::Common::Redis::RespType::Array);
```

Comment:
Should we have real error checking here? If not should these be RELEASE_ASSERTS? This is untrusted upstream data right?